dollarskip.c

```diff
@@ -1,36 +1,76 @@
-		if(strcmp(env, "")) { // if env != ""
-			strcpy(shell, env); // copy env to shell
-			strcat(shell, " -c \'"); // add " -c '"
-			strcat(command, shell); // add it to command
-			for (int i=1; i<argc; i++){
-				strcat(strcat(command,argv[i])," ");
-			} // add all the command (cmd args) to command
-			/****
-    			* remove the space from the end of the command
-    			* for example the command is "ls -l", it takes up 5 characters from the array (command variable) + a space at the end caused by the for loop above + a character at the end that is reserved for a NUL character.
-    			* in this case 'strlen(command)' will return 6, remove 1, and you get the location of the character in the array containing the extra space.
-    			* replace the space with NUL (null terminator), the end of the string.
-   	 		****/
-			command[strlen(command) - 1] = '\0';
-			strcat(command, "\'");// add "'"
-		} else { // if the SHELL env var doesn't exist or contain anything,
-			for (int i=1; i<argc; i++){ // put the command in 'command',
-				strcat(strcat(command,argv[i])," ");
-			}
-			command[strlen(command) - 1] = '\0'; // and remove the NULL space at the end.
+#include <assert.h>
+
+#ifndef COMMAND_MAX_SIZE
+#define COMMAND_MAX_SIZE 8096
+#endif
+
+/******
+ * Collect all the arguments (argv) excluding the first one (argv[0]) into a single string.
+ * returned string is heap allocated so you have to free() it.
+ * 
+ * @param argc argc from main()
+ * @param argv argv from main()
+ * @param argv_size pointer to an integer. if not NULL, the length of argv + 1 after each argument will be put in it.
+ * 
+ * @return A heap allocated string containing all of argv excluding argv[0]
+ ******/
+char *collect_args(int argc, char **argv, size_t *argv_size) {
+	size_t argv_len = 0;
+	char *args = NULL;
+
+	// get the size of all of argv + a space after each argument (last one for the nul character)
+	for(int i = 1; i < argc; ++i) {
+		argv_len += strlen(argv[i]) + 1; // 1 for the space and nul character at the end
+	}
+	// if argv_size isn't NULL, dereference it and put argv_len in it.
+	if(argv_size != NULL) *argv_size = argv_len;
+
+	// allocate memory for the argument string
+	args = malloc(argv_len);
+
+	// copy all of argv (exluding argv[0]) to args and add a space after each argument
+	for(int i = 1; i < argc; ++i) {
+		strcat(args, argv[i]);
+		strcat(args, " ");
+	}
+	// remove the space at the end
+	args[argv_len - 1] = '\0';
+	return args;
+}
+
+int main(int argc, char **argv) {
+	// if arguments are provided
+	if(argc > 1) {
+		char command[COMMAND_MAX_SIZE] = {0};
+		// get the default system shell
+		char *shell_env = getenv("SHELL");
+		// if it exists
+		if(shell_env != NULL && strcmp(shell_env, "")) {
+			size_t argv_len = 0;
+			// collect arguments
+			char *args = collect_args(argc, argv, &argv_len);
+			// check that there is enough space in 'command'
+			assert(argv_len + strlen(shell_env) + 6 < COMMAND_MAX_SIZE); // 6 = strlen(" -c ''")
```

Comment:
What does the assert do here?

Reply:
Check that there is enough space in 'command' to hold the command.

Reply:
:+1: